bazel/linting/defs.bzl

```diff
@@ -0,0 +1,190 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_context", "go_path")
+load("//bazel/utils:binary.bzl", "declare_binary", "download_binary")
+
+DepInfo = provider(
+    fields = {
+        "files": "deps",
+    },
+)
+
+def _dep_aspect_impl(target, ctx):
+    # Make sure the rule has a srcs attribute.
+    to_return = []
+    if hasattr(ctx.rule.attr, "srcs"):
+        for src in ctx.rule.attr.srcs:
+            for f in src.files.to_list():
+                to_return.append(f)
+
+    if hasattr(ctx.rule.attr, "deps"):
+        for dep in ctx.rule.attr.deps:
+            to_return.extend(dep[DepInfo].files)
+
+    return [DepInfo(files = to_return)]
+
+dep_aspect = aspect(
+    implementation = _dep_aspect_impl,
+    attr_aspects = ["deps"],
+)
+
+def _go_lint_impl(ctx):
+    go = go_context(ctx)
+    inputs = []
+    for l in ctx.attr.deps:
+        inputs.extend(l[DefaultInfo].files.to_list())
+
+    for lib in ctx.attr.go_libraries:
+        inputs.extend(lib.files.to_list())
+
+    outfiles = []
+    for index, target in enumerate(ctx.attr.targets):
+        lib = ctx.attr.go_libraries[index]
+        library_name = lib.label.name
+        outfile = ctx.actions.declare_file(target + ".txt")
+        outfiles.append(outfile)
+        ctx.actions.run(
+            inputs = depset(inputs),
+            outputs = [outfile],
+            progress_message = "Running Linter on GOPATH " + target,
+            executable = ctx.files._lint_script[0],
+            tools = [go.go, ctx.files._golangci_lint[0]],
+            env = {
+                "GO_LOCATION": go.sdk.root_file.dirname,
+                "GO_LIBRARY_NAME": library_name,
+                "LINT_OUTPUT": outfile.path,
+                "GIT_DATA": inputs[0].path,
+                "GOLANGCI_LINT": ctx.files._golangci_lint[0].path,
+                "TARGET": target,
+            },
+        )
+    return DefaultInfo(
+        files = depset(outfiles + ctx.files.deps),
+    )
+
+go_lint = rule(
```

Comment:
can you add `doc =` attributes to explain the rule, and the attributes?

eg, what is deps? list of .go files directly? When would you use it over `go_libraries`?